core/src/commonMain/kotlin/com/ustadmobile/core/viewmodel/report/ReportEditViewModel.kt

```diff
@@ -0,0 +1,230 @@
+
+class ReportEditViewModel(
+    di: DI,
+    savedStateHandle: UstadSavedStateHandle,
+    destName: String = ""
+) : UstadEditViewModel(di, savedStateHandle, DEST_NAME) {
+
+    private val _uiState: MutableStateFlow<ReportEditUiState> =
+        MutableStateFlow(ReportEditUiState())
+    val uiState: Flow<ReportEditUiState> = _uiState.asStateFlow()
+    private val entityUid: Long
+        get() = savedStateHandle[UstadView.ARG_ENTITY_UID]?.toLong() ?: 0
+
+    init {
+        loadingState = LoadingUiState.INDETERMINATE
+        val title = systemImpl.getString(MR.strings.edit_report)
+
+        _appUiState.update {
+            AppUiState(
+                title = title,
+                hideBottomNavigation = false
+            )
+        }
+        launchIfHasPermission(
+            setLoadingState = true,
+            permissionCheck = { true }
+        ) {
+            async {
+                loadEntity(
+                    serializer = Report.serializer(),
+                    onLoadFromDb = { db ->
+                        db.reportDao().findByUid(entityUid)
+                    },
+                    makeDefault = {
+                        Report()
+                    },
+                    uiUpdate = { loadedReport ->
+                        _uiState.update { prev ->
+                            prev.copy(
+                                reportOptions2 = loadedReport?.reportOptions?.takeIf { it.isNotBlank() }
+                                    ?.let { Json.decodeFromString(ReportOptions2.serializer(), it) }
+                                    ?: ReportOptions2()
+                            )
+                        }
+                    }
+                )
+                navResultReturner.filteredResultFlowForKey(RESULT_KEY_REPORT).collect { result ->
+                    val reportResult = result.result as? Map<*, *> ?: return@collect
+                    val filter = reportResult["filter"] as? ReportFilter3 ?: return@collect
+                    val seriesId = (reportResult["reportSeriesUid"] as? Int) ?: return@collect
+                    println("Report filter: ${filter}")
+                    onFilterChanged(filter, seriesId)
+                }
+            }
+        }
+
+        _appUiState.update { prev ->
+            prev.copy(
+                actionBarButtonState = ActionBarButtonUiState(
+                    visible = true,
+                    text = systemImpl.getString(MR.strings.done),
+                    onClick = this@ReportEditViewModel::onClickSave
+                )
+            )
+        }
+    }
+
+    fun onClickSave() {
+        viewModelScope.launch {
+            val currentReport = _uiState.value.reportOptions2
+            val report = Report(
+                reportTitle = currentReport.title,
+                reportOptions = Json.encodeToString(currentReport),
+            )
+            scheduleEntityCommitToSavedState(report, serializer = Report.serializer())
```

Comment:
scheduleEntityCommitToSavedState should be in the onEntityChanged function, and run almost immediately when the user changes anything.

The edit view models work as follows:

1) As the user modifies an entity through the UI form etc - those changes are put into the saved state handle. This way - any changes they made are saved in case the app is destroyed and recreated, and changes can be saved on the web if the user navigates backwards and forwards.

Changes are not (yet) saved in the database. The user might not want to save those changes, they can still discard those changes if they made mistakes etc.

2) Once the user clicks save, then the entity is saved into the database.